plugin/core/promise.py

```diff
@@ -184,3 +186,40 @@ def _is_resolved(self) -> bool:
     def _get_value(self) -> Any:
         with self.mutex:
             return self.value
+
+
+opening_files = {}  # type: Dict[str, Tuple[Promise, Callable[[sublime.View], None]]]
+
+
+def open_file(
+    window: sublime.Window,
+    file_path: str,
+    flags: int = 0,
+    group: int = -1
+) -> Promise:
+    """
+    Open a file asynchronously. It is only safe to call this function from the UI thread.
+    """
+
+    # Is the file already open?
+    view = window.find_open_file(file_path)
+    if view:
+        return Promise.resolve(view)
+
+    # Is the view opening right now? Then return the associated unresolved promise
+    for fn, value in opening_files.items():
+        if fn == file_path or os.path.samefile(fn, file_path):
+            return value[0]
+
+    # Prepare new promise
+    def fullfill(resolve: ResolveFunc) -> None:
+        global opening_files
+        opening_files[file_path] = (None, resolve)  # type: ignore
```

Comment:
So this reset the Promise to `None` when view is opened? What happens if that file is closed and then we get another open request? Won't the loop above just return None value then?

Reply:
The fullfill func is ran immediately upon creating the promise. It puts the resolve func in the dict, to be called later by on_load